src/app/websites/website-services/afterdark/afterdark.service.ts

```diff
@@ -0,0 +1,199 @@
+    if (!dotProp.get(responseBody, 'data.userViewer.user', null)) {
+      return null;
+    }
+    const collections = dotProp.get(responseBody, 'data.userViewer.user.collectionSet.edges', []);
+    return {
+      'userId': dotProp.get(responseBody, 'data.userViewer.user.id', ''),
+      'username': dotProp.get(responseBody, 'data.userViewer.user.username', ''),
+      'collections': collections.map(c => c.node),
+    };
+  }
+
+  public getFolders(profileId: string): Folder[] {
+    // conveniently, collections have the same structure as folders
+    return this._profileManager.getData(profileId, AfterDark.name).collections;
+  }
+
+  public async checkStatus(profileId: string, data?: any): Promise<WebsiteStatus> {
+    const returnValue: WebsiteStatus = {
+      username: null,
+      status: LoginStatus.LOGGED_OUT,
+    };
+    const csrfToken = await this.getCsrfToken(profileId);
+    const userData = await this.getUserData(profileId, csrfToken);
+    if (!userData) {
+      return returnValue;
+    }
+    this._profileManager.storeData(profileId, AfterDark.name, userData);
+    returnValue.username = userData.username;
+    returnValue.status = LoginStatus.LOGGED_IN;
+    return returnValue;
+  }
+
+  public async post(submission: Submission, postData: SubmissionPostData): Promise<PostResult> {
+    const cookies = await getCookies(postData.profileId, this.BASE_URL);
+    const csrfToken = await this.getCsrfToken(postData.profileId);
+    const fileFormData = fileAsFormDataObject(postData.primary);
+    const query = `mutation UploadImageMutation(
+      $filename: String!,
+      $tags: [String],
+      $collectionIds: [ID],
+      $title: String,
+      $artistUrl: String,
+      $artistName: String,
+      $nsfw: Boolean = true,
+      $claimedByUser: Boolean
+    ) {
+      uploadImage(
+        filename: $filename,
+        tags: $tags,
+        collectionIds: $collectionIds,
+        title: $title,
+        artistName: $artistName,
+        artistUrl: $artistUrl,
+        nsfw: $nsfw,
+        claimedByUser: $claimedByUser
+      ) {
+        ok
+        image {
+          id
+        }
+      }
+    }`;
+    const variables = JSON.stringify({
+      "filename": postData.primary.fileInfo.name,
+      "tags": postData.tags,
+      "collectionIds": postData.options.folders,
+      "title": postData.title,
+      "artistName": this._profileManager.getData(postData.profileId, AfterDark.name).username,
+      "artistUrl": "",
+      "nsfw": postData.rating !== SubmissionRating.GENERAL && postData.rating !== SubmissionRating.MATURE,
+      "claimedByUser": true,
+    });
+    let formData = {
+      query,
+      variables,
+    };
+    formData[postData.primary.fileInfo.name] = fileFormData;
+    const response = await got.post(`${this.BASE_URL}/graphql/`, formData, this.BASE_URL, cookies, {
+      headers: {
+        'content-type': 'multipart/form-data',
+        'X-CSRFToken': csrfToken,
+      }
+    });
```

Comment:
I thought an earlier commit said that graphql didn't accept multipart and only took json?
If it works its works, I just wanted to re-verify.

Reply:
The site is a bit weird - for posting an upload, it uses some hack to allow the large image body that involves taking form data, of which the usual graphql fields are form data elements.

An example taken directly from `afterdark.art/upload`: https://pastebin.com/raw/uxTGCGZQ

And result:
![2020-09-08-205816_833x805_scrot](https://user-images.githubusercontent.com/1375517/92552666-0e386280-f216-11ea-89e9-22fdbd684ead.png)